Source/Flow/Private/FlowAsset.cpp

```diff
@@ -421,34 +470,76 @@ void UFlowAsset::FinishFlow(const EFlowFinishPolicy InFinishPolicy, const bool b
 	}
 }
 
+bool UFlowAsset::HasStartedFlow() const
+{
+	return RecordedNodes.Num() > 0;
+}
+
+AActor* UFlowAsset::TryFindActorOwner() const
+{
+	UActorComponent* OwnerAsComponent = Cast<UActorComponent>(GetOwner());
+	if (IsValid(OwnerAsComponent))
+	{
+		return Cast<AActor>(OwnerAsComponent->GetOwner());
+	}
+
+	return nullptr;
+}
+
 TWeakObjectPtr<UFlowAsset> UFlowAsset::GetFlowInstance(UFlowNode_SubGraph* SubGraphNode) const
 {
 	return ActiveSubGraphs.FindRef(SubGraphNode);
 }
 
-void UFlowAsset::TriggerCustomEvent(UFlowNode_SubGraph* Node, const FName& EventName) const
+void UFlowAsset::TriggerSubgraphCustomInput(UFlowNode_SubGraph& Node, const FName& EventName) const
 {
-	const TWeakObjectPtr<UFlowAsset> FlowInstance = ActiveSubGraphs.FindRef(Node);
+	check(HasStartedFlow());
+
+	const TWeakObjectPtr<UFlowAsset> FlowInstance = ActiveSubGraphs.FindRef(&Node);
 	if (FlowInstance.IsValid())
 	{
-		for (UFlowNode_CustomInput* CustomInput : FlowInstance->CustomInputNodes)
+		FlowInstance->TriggerCustomInput(EventName);
+	}
+}
+
+void UFlowAsset::TriggerCustomInput(const FName& EventName)
+{
+	check(HasStartedFlow());
+
+	for (UFlowNode_CustomInput* CustomInput : CustomInputNodes)
+	{
+		if (CustomInput->EventName == EventName)
 		{
-			if (CustomInput->EventName == EventName)
-			{
-				FlowInstance->RecordedNodes.Add(CustomInput);
-				CustomInput->TriggerFirstOutput(true);
-			}
+			RecordedNodes.Add(CustomInput);
+
+			CustomInput->ExecuteInput(EventName);
 		}
 	}
 }
 
-void UFlowAsset::TriggerCustomOutput(const FName& EventName) const
+void UFlowAsset::TriggerCustomOutput(const FName& EventName)
 {
-	NodeOwningThisAssetInstance->TriggerOutput(EventName);
+	check(HasStartedFlow());
+
+	const bool bIsSubgraph = NodeOwningThisAssetInstance.IsValid();
+	if (bIsSubgraph)
```

Comment:
Here is the inflection point where we change where the output trigger information goes to.

For Subgraphs, it does the previous behavior.

For root graphs it dispatches the event to the owning component.

Reply:
Neat!
Only tweaking comments here :)